spark_log_parser/parsing_models/application_model.py

```diff
@@ -329,16 +316,44 @@ def __init__(self, eventlogpath, bucket=None, stdoutpath=None, debug=False):  #
         self.num_instances = len(numpy.unique(hosts))
         self.executors_per_instance = numpy.ceil(self.num_executors / self.num_instances)
 
-        # print("Finished reading input data:")
+        for task in self.tasks:
+            stage_id = task.stage_id
+            stage = self.stages[stage_id]
+            stage.add_task(task)
+
+        for stage_id, stage in self.stages.items():
+            if stage.submission_time is not None:
+                raise UrgentEventValidationException(missing_event=f"Stage {stage_id} Submit")
+
+            job_ids_for_stage = self.jobs_for_stage.get(stage_id)
+            if (not job_ids_for_stage) and (not debug):
+                # If this stage is not associated with any particular job, that likely means we are missing some data
+                raise UrgentEventValidationException(
+                    missing_event=f"Job Start for Stage {stage_id}"
+                )
+
+            for job_id in job_ids_for_stage:
+                self.jobs[job_id].stages[stage_id] = stage
+
+            stage.finalize_tasks()
+
         for job_id, job in self.jobs.items():
             job.initialize_job()
-            # print("Job", job_id, " has stages: ", job.stages.keys())
 
     def output_all_job_info(self):
         for job_id, job in self.jobs.items():
             features_filename = f"{os.path.dirname(self.eventlogpath)}/e{self.num_executors}_p{self.parallelism}_mem{self.memory_per_executor}_job{job_id}"
             job.write_features(features_filename)
 
+    def maybe_set_new_finish_time(self, new_finish_time: int):
+        """
+        As we read log lines, we want to mark the 'latest' finish_time we have seen. This may come from a few different
+        events, since there is no single event that is guaranteed to be present in the log from which we can determine
+        the application's finish_time
+        """
+        if not self.finish_time or new_finish_time > self.finish_time:
+            self.finish_time = new_finish_time
```

Comment:
There are multiple events where we may get our "final" timestamp. Now that we are processing log lines out-of-order, that means that we want to make sure we only ever take the maximum of all the values we were considering.